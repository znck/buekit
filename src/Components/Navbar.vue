<script>
import  { isFunction } from 'lodash'
import PropTypes from '@znck/prop-types'

import { styleResolver, style, colors } from '../utils'

import Burger from './NavbarBurger.vue'

export default {
  name: 'Navbar',
  mixins: [colors],
  data: () => ({ active: false }),
  props: {
    /**
     * Background color of navbar.
     * @version 0.0.0
     * @since Version 0.0.0
     */
    color: PropTypes.oneOf('white', ...colors),
    /**
     * Navbar without burger
     * @version 0.0.0
     * @since Version 0.0.0
     */
    noBurger: PropTypes.bool.value(false),
    /**
     * Wrapping html tag
     * @version 0.0.0
     * @since Version 0.0.0
     */
    tag: PropTypes.string.value('div'),
    /**
     * With transparent background
     * @version 0.0.0
     * @since Version 0.0.0
     */
    transparent: PropTypes.bool.value(false)
  },
  render (h) {
    const _ = styleResolver(this.$style)
    const { color, transparent, tag, noBurger } = this.$props
    const active = this.active
    let { menu, default: brand = [], start: left, end: right } = this.$slots

    if (menu) { // menu goes to left.
      left = menu.concat(left || [])
    }

    const needsBrand = !noBurger || brand
    const needsMenu = left || right

    if (!noBurger) {
      brand.push(
        h(Burger, {
          class: active && _('is-active'),
          on: { click: () => this.activate(!active) }
        })
      )
    }

    return style(
      h(tag, { props: this.$attrs, on: this.$listeners }, [
        /* Navbar Brand */
        needsBrand && h('div', { class: _('navbar-brand') }, brand),
        /* Navbar Menu */
        needsMenu && h('div', { class: [_('navbar-menu'), active && _('is-active')] }, [
          left && h('div', { class: _('navbar-start') }, left),
          right && h('div', { class: _('navbar-end') }, right)
        ])
      ]),
      _('navbar'),
      color && _(`is-${color}`),
      transparent && _('is-transparent')
    )
  },

  methods: {
    activate (value) {
      this.active = value
    }
  }
}
</script>

<style lang="scss" module>
@import '~bulma/sass/utilities/_all.sass';
@import '../_base.scss';
@import '~bulma/sass/components/navbar.sass';
</style>
