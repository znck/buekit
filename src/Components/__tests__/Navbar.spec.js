import { shallow } from 'vue-test-utils'
import { createRenderer } from 'vue-server-renderer'

import Navbar from '../Navbar.vue'

const renderer = createRenderer()
function snapshots(...options) {
	options.forEach(
    option => it('matches snapshot: ' + JSON.stringify(option), () => {
      const { props: propsData, slots = {} } = option
      slots.default = slots.default || '<p>content</p>'
      const wrapper = shallow(Navbar, { propsData, slots })

      slots.default === '<p>content</p>' && // If default slot is not provied.
      expect(wrapper.contains('p')).toBe(true)

      renderer.renderToString(wrapper.vm, (err, str) => {
        if (err) {
          throw new Error(err)
        }

        expect(str).toMatchSnapshot()
      })
    })
  )
}

describe('Navbar.vue', () => {
  snapshots(
    {},
    {props: {color: 'white'}},
    {props: {noBurger: true}},
    {props: {tag: 'nav'}},
    {props: {transparent: true}},
    {slots: {start: '<p>content</p>'}},
    {slots: {end: '<p>content</p>'}},
    {slots: {menu: '<p>content</p>'}}
  )

  it('should have background', () => {
    const wrapper = shallow(Navbar, { propsData: { color: 'white' } })

    expect(wrapper.hasClass('is-white')).toBe(true)
  })

  it('should validate background color', () => {
    const wrapper = shallow(Navbar, { propsData: { color: 'any' } })

    expect(wrapper.vm.$options.props.color.validator('any')).toBe(false)
  })
})
