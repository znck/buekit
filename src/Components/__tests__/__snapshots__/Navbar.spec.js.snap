// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar.vue matches snapshot: {"props":{"color":"white"}} 1`] = `
<div class="navbar is-white">
    <div class="navbar-brand">
        <p>content</p><button class="button navbar-burger"><span></span><span></span><span></span></button><button class="button navbar-burger"><span></span><span></span><span></span></button></div>
</div>
`;

exports[`Navbar.vue matches snapshot: {"props":{"noBurger":true}} 1`] = `
<div class="navbar">
    <div class="navbar-brand">
        <p>content</p>
    </div>
</div>
`;

exports[`Navbar.vue matches snapshot: {"props":{"tag":"nav"}} 1`] = `
<nav class="navbar">
    <div class="navbar-brand">
        <p>content</p><button class="button navbar-burger"><span></span><span></span><span></span></button><button class="button navbar-burger"><span></span><span></span><span></span></button></div>
</nav>
`;

exports[`Navbar.vue matches snapshot: {"props":{"transparent":true}} 1`] = `
<div class="navbar is-transparent">
    <div class="navbar-brand">
        <p>content</p><button class="button navbar-burger"><span></span><span></span><span></span></button><button class="button navbar-burger"><span></span><span></span><span></span></button></div>
</div>
`;

exports[`Navbar.vue matches snapshot: {"slots":{"end":"<p>content</p>"}} 1`] = `
<div class="navbar">
    <div class="navbar-brand">
        <p>content</p><button class="button navbar-burger"><span></span><span></span><span></span></button><button class="button navbar-burger"><span></span><span></span><span></span></button></div>
    <div class="navbar-menu">
        <div class="navbar-end">
            <p>content</p>
        </div>
    </div>
</div>
`;

exports[`Navbar.vue matches snapshot: {"slots":{"menu":"<p>content</p>"}} 1`] = `
<div class="navbar">
    <div class="navbar-brand">
        <p>content</p><button class="button navbar-burger"><span></span><span></span><span></span></button><button class="button navbar-burger"><span></span><span></span><span></span></button></div>
    <div class="navbar-menu">
        <div class="navbar-start">
            <p>content</p>
        </div>
    </div>
</div>
`;

exports[`Navbar.vue matches snapshot: {"slots":{"start":"<p>content</p>"}} 1`] = `
<div class="navbar">
    <div class="navbar-brand">
        <p>content</p><button class="button navbar-burger"><span></span><span></span><span></span></button><button class="button navbar-burger"><span></span><span></span><span></span></button></div>
    <div class="navbar-menu">
        <div class="navbar-start">
            <p>content</p>
        </div>
    </div>
</div>
`;

exports[`Navbar.vue matches snapshot: {} 1`] = `
<div class="navbar">
    <div class="navbar-brand">
        <p>content</p><button class="button navbar-burger"><span></span><span></span><span></span></button><button class="button navbar-burger"><span></span><span></span><span></span></button></div>
</div>
`;
